R/core.R

```diff
@@ -0,0 +1,363 @@
+      system_parameters
+    )
+
+    without_nulls <- rm_null_obs(with_all_fields)
+  } else {
+    without_nulls <- rm_null_obs(system_parameters)
+  }
+
+  # rm_null_obs transforms everything into a list, and we need
+  # the stacktrace as a data.frame/tibble
+  without_nulls$exception$stacktrace <- system_parameters$exception$stacktrace
+
+  payload <- jsonlite::toJSON(
+    without_nulls,
+    auto_unbox = TRUE,
+    null = "null",
+    na = "null"
+  )
+
+  return(payload)
+}
+
+
+#' Send a message to a Sentry server
+#'
+#' @param ... named parameters
+#'
+#' @return sends message to Sentry
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' capture(message = "oh hai there!") # send message to sentry
+#' }
+capture <- function(...) {
+  browser()
+  if (!is_sentry_configured()) {
+    stop("Sentry is not configured!")
+  }
+
+  payload <- prepare_payload(...)
+
+  resp <- httr::POST(
+    url = sentry_url(),
+    body = payload, encode = "json",
+    httr::add_headers(.headers = sentry_headers())
+  )
+
+  resp_status_code <- httr::status_code(resp)
+
+  if (resp_status_code == 201 || resp_status_code == 200) {
+    resp_body <- httr::content(resp)
+    message("Your event was recorder in Sentry with ID ", resp_body$id)
+  } else {
+    warning(
+      "Error connecting to Sentry:",
+      httr::content(resp, "text", encoding = "UTF-8")
+    )
+  }
+}
+
+
+#' Report a message to Sentry
+#'
+#' @param message message text
+#' @param ... optional additional named parameters
+#' @param level the level of the message. Default: "info"
+#'
+#' @return nothing; sends message to Sentry
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' capture_message("this is an important message", logger = "my.logger")
+#' }
+capture_message <- function(message, ..., level = "info") {
+  # TODO: hello? happy path?
+  capture(message = message, ..., level = level)
+}
+
+
+#' Report an error or exception object
+#'
+#' @param error an error object
+#' @param ... optional additional named parameters
+#' @param level the level of the message. Default: "error"
+#'
+#' @return nothing; sends error to Sentry
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' capture_exception(simpleError("foo"), tags = list(version = "1.0"))
+#' }
+capture_exception <- function(error, ..., level = "error") {
+  # TODO: hello? happy path?
+  # TODO: wrangling the error should be a pure function prepare_error()
+
+  if ("function_calls" %in% names(error)) {
+    stacktrace <- calls_to_stacktrace(error$function_calls)
+  } else {
+    stacktrace <- calls_to_stacktrace(sys.calls())
+  }
+
+  error_type <- class(error)[[1]]
+  error_message <- gsub('(\\n|\\")', "", as.character(error))
+
+  # tibble allows list-columns, which jsonlite translate to array of maps
+  exception_payload <- list(
+    type = error_type,
+    value = error_message,
+    stacktrace = list(
+      frames = stacktrace
+    )
+  )
+
+  capture(exception = exception_payload, ..., level = level)
+}
+
+
+#' Build the response URL
```

Comment:
Response?

Reply:
@kirel what do you think about `Build the sentry.io call URL`?